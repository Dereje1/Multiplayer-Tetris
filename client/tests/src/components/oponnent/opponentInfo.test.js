/**
 * @jest-environment jsdom
 */
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import OpponentInfo from '../../../../src/components/oponnent/opponentInfo';
import { gameStub } from '../../../stub'

describe('The opponent info', () => {
    let props, windowSpy;
    let assign = jest.fn();
    beforeEach(() => {
        props = {
            socketState: {},
            difficulty: 2,
            setDifficulty: jest.fn(),
            requestInvite: jest.fn(),
            getPool: jest.fn(),
            authenticated: true
        }
        windowSpy = jest.spyOn(global, "window", "get");
        windowSpy.mockImplementation(() => ({ location: { assign } }));
    })
    afterEach(() => {
        props = null
        windowSpy.mockClear()
    })

    test('will render login for unauthenticated users', async () => {
        const updatedProps = {
            ...props,
            authenticated: false
        }
        const wrapper = shallow(<OpponentInfo {...updatedProps} />)
        expect(toJson(wrapper)).toMatchSnapshot();
        const loginButton = wrapper.find({id: 'googleloginbutton'})
        loginButton.props().onClick()
        expect(assign).toHaveBeenCalledWith('/auth/google')
    })

    test('will render stage 1', () => {
        // stage 1 - no logged in opponents in multiplayer mode found
        const updatedProps = {
            ...props,
            socketState: {
                opponents: []
            }
        }
        const wrapper = shallow(<OpponentInfo {...updatedProps} />);
        expect(toJson(wrapper)).toMatchSnapshot();
        const reload = wrapper.find({ className: 'opponentContainer__opponentDescription__invitation_reload' })
        reload.props().onClick()
        expect(props.getPool).toHaveBeenCalledTimes(1)
    })

    test('will render stage 2', () => {
        // stage 2 - logged in opponents in multiplayer mode found, display invitation buttons
        const updatedProps = {
            ...props,
            socketState: {
                opponents: [{
                    displayname: "stub opponent 1",
                    socketId: "stub-opponent-1-socketID"
                }, {
                    displayname: "stub opponent 2",
                    socketId: "stub-opponent-2-socketID"
                }]
            }
        }
        const wrapper = shallow(<OpponentInfo {...updatedProps} />);
        expect(toJson(wrapper)).toMatchSnapshot();
        const difficultyButton = wrapper.find({ role: 'button' }).at(2)
        const inviteButton = wrapper.find({ type: 'submit' }).at(0)
        difficultyButton.props().onClick()
        inviteButton.props().onClick()
        expect(props.setDifficulty).toHaveBeenCalledWith(3)
        expect(props.requestInvite).toHaveBeenCalledWith('stub-opponent-1-socketID')
    })

    test('will render stage 3A', () => {
        // stage 3A - display pending invitation sent
        const updatedProps = {
            ...props,
            socketState: {
                invitationTo: {
                    displaynameReciever: "stub opponent 1",
                    socketIdReciever: "stub-opponent-1-socketID"
                }
            }
        }
        const wrapper = shallow(<OpponentInfo {...updatedProps} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    })

    test('will render stage 3B', () => {
        // stage 3B - display declined invitation
        const updatedProps = {
            ...props,
            socketState: {
                declinedInvitation: {}
            }
        }
        const wrapper = shallow(<OpponentInfo {...updatedProps} />);
        const reload = wrapper.find({ className: 'opponentContainer__opponentDescription__invitation_reload' })
        expect(toJson(wrapper)).toMatchSnapshot();
        reload.props().onClick()
        expect(props.getPool).toHaveBeenCalledTimes(1)
    })

    test('will render stage 4', () => {
        // stage 4 - invitation has been accepted, display pre game warm up
        const updatedProps = {
            ...props,
            socketState: {
                acceptedInvitation: {
                    opponnetDisplayname: 'stub opponent 1',
                    countdown: 7
                }
            }
        }
        const wrapper = shallow(<OpponentInfo {...updatedProps} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    })

    test('will render stage 5', () => {
        // stage 5 - Game has started
        const updatedProps = {
            ...props,
            socketState: {
                gameInProgress: {
                    opponentScreen: JSON.stringify(gameStub),
                    info: {
                        opponnetDisplayname: 'stub opponent 1'
                    }
                }
            }
        }
        const wrapper = shallow(<OpponentInfo {...updatedProps} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    })

    test('will render stage 6', () => {
        // stage 6 - Game over
        const updatedProps = {
            ...props,
            socketState: {
                gameOver: {}
            }
        }
        const wrapper = shallow(<OpponentInfo {...updatedProps} />);
        expect(toJson(wrapper)).toMatchSnapshot();
        const reload = wrapper.find({ className: 'opponentContainer__opponentDescription__invitation_reload' })
        reload.props().onClick()
        expect(props.getPool).toHaveBeenCalledTimes(1)
    })

    test('will render loading wheel for all other states', () => {
        const wrapper = shallow(<OpponentInfo {...props} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    })
})

