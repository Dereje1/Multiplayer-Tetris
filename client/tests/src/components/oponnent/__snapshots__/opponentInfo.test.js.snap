// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The opponent info will render loading wheel for all other states 1`] = `
<div
  className="opponentContainer__opponentDescription"
>
  <ForwardRef(CircularProgress)
    style={
      Object {
        "marginTop": 80,
      }
    }
  />
</div>
`;

exports[`The opponent info will render login for unauthenticated users 1`] = `
<ForwardRef(Button)
  id="googleloginbutton"
  onClick={[Function]}
  startIcon={
    <Memo
      style={
        Object {
          "fontSize": 25,
        }
      }
    />
  }
  variant="outlined"
>
  Continue With Google
</ForwardRef(Button)>
`;

exports[`The opponent info will render stage 1 1`] = `
<div
  className="opponentContainer__opponentDescription"
>
  <p
    className="writing"
  >
    No opponents  
  </p>
  <FontAwesomeIcon
    beat={false}
    beatFade={false}
    border={false}
    bounce={false}
    className="opponentContainer__opponentDescription__invitation_reload"
    fade={false}
    fixedWidth={false}
    flip={false}
    icon={
      Object {
        "icon": Array [
          512,
          512,
          Array [],
          "f2f1",
          "M370.72 133.28C339.458 104.008 298.888 87.962 255.848 88c-77.458.068-144.328 53.178-162.791 126.85-1.344 5.363-6.122 9.15-11.651 9.15H24.103c-7.498 0-13.194-6.807-11.807-14.176C33.933 94.924 134.813 8 256 8c66.448 0 126.791 26.136 171.315 68.685L463.03 40.97C478.149 25.851 504 36.559 504 57.941V192c0 13.255-10.745 24-24 24H345.941c-21.382 0-32.09-25.851-16.971-40.971l41.75-41.749zM32 296h134.059c21.382 0 32.09 25.851 16.971 40.971l-41.75 41.75c31.262 29.273 71.835 45.319 114.876 45.28 77.418-.07 144.315-53.144 162.787-126.849 1.344-5.363 6.122-9.15 11.651-9.15h57.304c7.498 0 13.194 6.807 11.807 14.176C478.067 417.076 377.187 504 256 504c-66.448 0-126.791-26.136-171.315-68.685L48.97 471.03C33.851 486.149 8 475.441 8 454.059V320c0-13.255 10.745-24 24-24z",
        ],
        "iconName": "sync-alt",
        "prefix": "fas",
      }
    }
    inverse={false}
    listItem={false}
    mask={null}
    maskId={null}
    onClick={[Function]}
    pull={null}
    pulse={false}
    rotation={null}
    shake={false}
    size={null}
    spin={false}
    spinPulse={false}
    spinReverse={false}
    swapOpacity={false}
    symbol={false}
    title=""
    titleId={null}
    transform={null}
  />
</div>
`;

exports[`The opponent info will render stage 2 1`] = `
<ForwardRef(Box)
  className="opponentContainer__opponentDescription"
>
  <p
    className="writing"
  >
    Difficulty
  </p>
  <ForwardRef(ButtonGroup)
    aria-label="outlined primary button group"
    color="primary"
  >
    <ForwardRef(Button)
      key="1"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      size="small"
      style={
        Object {
          "maxHeight": 35,
          "maxWidth": 35,
          "minHeight": 35,
          "minWidth": 35,
        }
      }
      tabIndex={0}
      variant="outlined"
    >
      1
    </ForwardRef(Button)>
    <ForwardRef(Button)
      key="2"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      size="small"
      style={
        Object {
          "maxHeight": 35,
          "maxWidth": 35,
          "minHeight": 35,
          "minWidth": 35,
        }
      }
      tabIndex={0}
      variant="contained"
    >
      2
    </ForwardRef(Button)>
    <ForwardRef(Button)
      key="3"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      size="small"
      style={
        Object {
          "maxHeight": 35,
          "maxWidth": 35,
          "minHeight": 35,
          "minWidth": 35,
        }
      }
      tabIndex={0}
      variant="outlined"
    >
      3
    </ForwardRef(Button)>
    <ForwardRef(Button)
      key="4"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      size="small"
      style={
        Object {
          "maxHeight": 35,
          "maxWidth": 35,
          "minHeight": 35,
          "minWidth": 35,
        }
      }
      tabIndex={0}
      variant="outlined"
    >
      4
    </ForwardRef(Button)>
  </ForwardRef(ButtonGroup)>
  <p
    className="writing"
  >
    Invite
  </p>
  <ForwardRef(ButtonGroup)
    aria-label="vertical outlined primary button group"
    color="primary"
    orientation="vertical"
  >
    <ForwardRef(Button)
      key="stub-opponent-1-socketID"
      onClick={[Function]}
      size="large"
      type="submit"
      variant="contained"
    >
      stub
    </ForwardRef(Button)>
    <ForwardRef(Button)
      key="stub-opponent-2-socketID"
      onClick={[Function]}
      size="large"
      type="submit"
      variant="contained"
    >
      stub
    </ForwardRef(Button)>
  </ForwardRef(ButtonGroup)>
</ForwardRef(Box)>
`;

exports[`The opponent info will render stage 3A 1`] = `
<div
  className="opponentContainer__opponentDescription"
>
  <div
    className="opponentContainer__opponentDescription__invitation"
  >
    <p
      className="writing"
    >
      Pending
    </p>
    <p
      className="writing"
    >
      Invitation to
    </p>
    <p
      className="writing"
    >
      stub
    </p>
    <ForwardRef(CircularProgress) />
  </div>
</div>
`;

exports[`The opponent info will render stage 3B 1`] = `
<div
  className="opponentContainer__opponentDescription"
>
  <div
    className="opponentContainer__opponentDescription__invitation"
  >
    <p
      className="writing"
    >
      Invitation
    </p>
    <p
      className="writing"
    >
      Declined
    </p>
    <FontAwesomeIcon
      beat={false}
      beatFade={false}
      border={false}
      bounce={false}
      className="opponentContainer__opponentDescription__invitation_reload"
      fade={false}
      fixedWidth={false}
      flip={false}
      icon={
        Object {
          "icon": Array [
            512,
            512,
            Array [],
            "f2f1",
            "M370.72 133.28C339.458 104.008 298.888 87.962 255.848 88c-77.458.068-144.328 53.178-162.791 126.85-1.344 5.363-6.122 9.15-11.651 9.15H24.103c-7.498 0-13.194-6.807-11.807-14.176C33.933 94.924 134.813 8 256 8c66.448 0 126.791 26.136 171.315 68.685L463.03 40.97C478.149 25.851 504 36.559 504 57.941V192c0 13.255-10.745 24-24 24H345.941c-21.382 0-32.09-25.851-16.971-40.971l41.75-41.749zM32 296h134.059c21.382 0 32.09 25.851 16.971 40.971l-41.75 41.75c31.262 29.273 71.835 45.319 114.876 45.28 77.418-.07 144.315-53.144 162.787-126.849 1.344-5.363 6.122-9.15 11.651-9.15h57.304c7.498 0 13.194 6.807 11.807 14.176C478.067 417.076 377.187 504 256 504c-66.448 0-126.791-26.136-171.315-68.685L48.97 471.03C33.851 486.149 8 475.441 8 454.059V320c0-13.255 10.745-24 24-24z",
          ],
          "iconName": "sync-alt",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      maskId={null}
      onClick={[Function]}
      pull={null}
      pulse={false}
      rotation={null}
      shake={false}
      size={null}
      spin={false}
      spinPulse={false}
      spinReverse={false}
      swapOpacity={false}
      symbol={false}
      title=""
      titleId={null}
      transform={null}
    />
  </div>
</div>
`;

exports[`The opponent info will render stage 4 1`] = `
<div
  className="opponentContainer__opponentDescription"
>
  <div
    className="opponentContainer__opponentDescription__Timer"
  >
    <h4>
      GET READY
    </h4>
    <h4>
      TO DUEL WITH:
    </h4>
    <p
      className="timercountdown"
    >
      stub
    </p>
    <p
      className="timercountdown"
    >
      in 7 s
    </p>
  </div>
</div>
`;

exports[`The opponent info will render stage 5 1`] = `
<div
  className="opponentContainer__opponentDescription"
>
  <div
    className="opponentContainer__opponentDescription__GamePlay"
  >
    <p
      className="writing"
    >
      stub
    </p>
    <p
      className="writing"
    >
      Lines Cleared
    </p>
    <p
      className="opponentContainer__opponentDescription__GamePlay__linescleared"
    >
      0
    </p>
  </div>
</div>
`;

exports[`The opponent info will render stage 6 1`] = `
<div
  className="opponentContainer__opponentDescription"
>
  <div
    className="opponentContainer__opponentDescription__invitation"
  >
    <FontAwesomeIcon
      beat={false}
      beatFade={false}
      border={false}
      bounce={false}
      className="opponentContainer__opponentDescription__invitation_reload"
      fade={false}
      fixedWidth={false}
      flip={false}
      icon={
        Object {
          "icon": Array [
            512,
            512,
            Array [],
            "f2f1",
            "M370.72 133.28C339.458 104.008 298.888 87.962 255.848 88c-77.458.068-144.328 53.178-162.791 126.85-1.344 5.363-6.122 9.15-11.651 9.15H24.103c-7.498 0-13.194-6.807-11.807-14.176C33.933 94.924 134.813 8 256 8c66.448 0 126.791 26.136 171.315 68.685L463.03 40.97C478.149 25.851 504 36.559 504 57.941V192c0 13.255-10.745 24-24 24H345.941c-21.382 0-32.09-25.851-16.971-40.971l41.75-41.749zM32 296h134.059c21.382 0 32.09 25.851 16.971 40.971l-41.75 41.75c31.262 29.273 71.835 45.319 114.876 45.28 77.418-.07 144.315-53.144 162.787-126.849 1.344-5.363 6.122-9.15 11.651-9.15h57.304c7.498 0 13.194 6.807 11.807 14.176C478.067 417.076 377.187 504 256 504c-66.448 0-126.791-26.136-171.315-68.685L48.97 471.03C33.851 486.149 8 475.441 8 454.059V320c0-13.255 10.745-24 24-24z",
          ],
          "iconName": "sync-alt",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      maskId={null}
      onClick={[Function]}
      pull={null}
      pulse={false}
      rotation={null}
      shake={false}
      size={null}
      spin={false}
      spinPulse={false}
      spinReverse={false}
      swapOpacity={false}
      symbol={false}
      title=""
      titleId={null}
      transform={null}
    />
  </div>
</div>
`;
